### PR TITLE
Handle query cancellation for waiting clients

### PR DESCRIPTION
**Pull request by mugunthan7285**
This fix enables the cancellation of waiting clients. Previously, if a cancel request arrived for a waiting client, it was ignored, causing the client to hang. With this fix, the behavior now aligns with how cancel requests are handled on a PostgreSQL server

### CODE REVIEW

File: src/client.c [hunk 1]
```diff
@@ -1367,6 +1367,32 @@ static bool handle_client_work(PgSocket *client, PktHdr *pkt)
...
+			client->query_start = 0;
+			client->xact_start = 0;
+			client->wait_for_cancel = false;
+			break;
+		default:
+			return false;
+			break;
+		}
+
+		sbuf_prepare_skip(sbuf, pkt->len);
```
Comment (JelteF) [inline src/client.c:1385]: This is not enough in case of the extended protocol. In that case we need to consume all messages until the next `Sync`. You should look at some of the pipeline tests that we have and test that everything is popped from the queue.

**Review by JelteF — CHANGES_REQUESTED** [review]
I think this is a useful change, but the current implementation is not handling some important edge cases afaict.

File: src/client.c (hunk 1 above)
Comment (mugunthan7285) [reply src/client.c:1385]: Hi @JelteF , Thanks for the review. I’m making the necessary corrections. Meanwhile, during testing, I had another concern. Since we first send an error response before processing the requests and if it is an explicit transaction the client might initiate rollback(). In that case, I believe we also need to handle it; otherwise, those packets could remain pending, waiting for the server connection again.

While testing this with psycopg, I observed that `conn.rollback()` did not send a `ROLLBACK` command because it checks the current transaction status first. However, if some implementations send a ROLLBACK blindly without performing the transaction status check, it could end up being queued. I’m unsure if my understanding is correct—your insights would be helpful.

Comment (JelteF) [reply src/client.c:1385]: I think sending the rollback to the actual postgres server is not a huge problem. At that point there's no transaction so a rollback is a no-op. I think a bigger problem is when the next Query packet is actually something that was supposed to be part of the transaction. So for instance Pgbouncer receives messages in this order (while no server is available)

1. Query("BEGIN")
2. Query("INSERT INTO t VALUES (1)")
3. CancelRequest()
4. Query("ROLLBACK")

I'm not sure what postgres behaviour is if a "BEGIN" gets cancelled, but I think that should be investigated before we can merge this. Because I think in the approach suggested so far. We would now throw an error for the BEGIN, but then the INSERT would go through to postgres like normal, i.e. *outside of a transaction*. And then the ROLLBACK would do nothing.

**Comment by AndrewJackson2020** [comment]
Just want to make sure there is a link to the originating issue with this PR.

https://github.com/pgbouncer/pgbouncer/issues/1222

**Comment by JelteF** [comment]
@mugunthan7285 do you intend to continue working on this?
